### ISSUE TITLE
Mode Changes Not Updating NeoPixels

### ISSUE THREAD

**Issue by CerberusVelvet** [prickle/Mavpixel#1]
Using - APM:Copter 3.5.0-rc7

In FM Stabilise neopixels set to 'Modes & Orientation'. Pixels light and colour can be adjusted in 'Mode Colors' and neopixel colours change.

Change to any other FM, pixels do not change as expected. 

In CLI using 'VARS' command mode number is changing. But pixels not updating.

Possibly missing code to update pixel status on FM change?

**Comment by CerberusVelvet** [comment]
ledstrip.ino, line 536
        applyDirectionalModeColor(ledIndex, ledConfig, flMode);

Changed to -
        applyDirectionalModeColor(ledIndex, ledConfig, iob_mode);

**Comment by Cerberus74** [comment]
Discovered the flight mode defines. Therefore for consistency should be using 'flMode' not 'iob_mode'.

**Comment by prickle** [comment]
I would expect that modification to produce some colourful but somewhat unpredictable results.
I think it should work as is, did you remember to assign compass directions to your navigation lights in the grid as well as in 'Mode Colours'?

**Comment by CerberusVelvet** [comment]
Hi,

I can confirm definitely does not currently work 'as is'. I assigned all LED's as 'down'. Changing modes on the PX4 results in no change at the neopixels. If I change the colour associated with flight mode '0' (stabilise), and direction 'down' the appropriate neopixels stay that colour no matter what mode is engaged. So the code is always passing a 0 to applyDirectionalModeColor I think.

I've not completely gotten to grips with the code yet so I certainly defer to your greater knowledge :)

Perhaps something is not working as expected in CheckFlightMode() ?

**Comment by prickle** [comment]
Hi Cerbs. 
Ah well, worth a shot. Sorry, I haven't played with this code for a while and don't have a build environment set up right now. 

Hmm. I guess you have the second serial set up. Try putting `dbSerial.println(flMode);` before `applyDirectionalModeColor()` in `applyLedModeLayer()` and see if anything prints in the CLI.

So when `iob_mode` is picked up (Mavlink.ino, 158) the function `CheckFlightMode()` (IOBoard_Funcs.ino, 118) is called soon after to set `flMode` to the internal mode enumeration. `applyLedModeLayer()` then uses this value only on leds that are flagged `LED_FUNCTION_FLIGHT_MODE`. Let's see if we get that far.
Good luck,
Nick.

**Comment by CerberusVelvet** [comment]
My line of thought went to iob_modes as when the 'vars' command is sent it dumps

 out(F("Modes:"), stream);
 outln(iob_mode, stream);

Which made me think iob_modes was the correct var. Until I looked closer and found the internal mode enumeration. I see now that really we want flModes doing the work. Makes sense.

So, I've added

 out(F("FLMode:"), stream);
 outln(flMode, stream);

To #ifdef DUMPVARS. But haven't got a pixhawk nearby me at the moment to test :(

I will also add the dbSerial.println(flMode); as you suggested and let you know what it says as soon as I get a chance.

Looking at CheckFlightMode() do you think it is possible that the newest copter version could be returning a value other than 1 or 2 from mavlink_msg_heartbeat_get_type(&msg); Which could cause the CheckFlightMode() to misbehave?

I'm guessing..... I should just wait and test it when I have a pixhawk to hand. 

Great code by the way. I'm very grateful to you for bringing LED's to the PX4!

Cheers!
Cerbs

**Comment by CerberusVelvet** [comment]
Soooooo I got curious and had a look at "mavlink_msg_heartbeat_get_type"

and I think I have figured it out..... maybe....

CheckFlightMode() is checking if apm_mav_type is "2" or "1" which from the mavlink common message set (https://pixhawk.ethz.ch/mavlink/) means is

1 - mav_type_fixed_wing
or
2 - mav_type_quadrotor

But.... my vehicle is a hexacopter! So I am betting my apm_mav_type is 

13 - mav_type_hexarotor

which is causing CheckFlightMode() to drop out with no pattern matched!

I will add some debug to report the apm_mav_type to confirm the theory... but it fits the behaviour?

**Comment by prickle** [comment]
Yes I see. Whoops, somewhat of an omission, I forgot your vehicle type!

I guess I will need to add at least vehicle types:
3	MAV_TYPE_COAXIAL	Coaxial helicopter
4	MAV_TYPE_HELICOPTER	Normal helicopter with tail rotor.
13	MAV_TYPE_HEXAROTOR	Hexarotor
14	MAV_TYPE_OCTOROTOR	Octorotor
15	MAV_TYPE_TRICOPTER	Tricopter

to multirotor, and at least types:

19	MAV_TYPE_VTOL_DUOROTOR	Two-rotor VTOL using control surfaces in vertical operation in addition. Tailsitter.
20	MAV_TYPE_VTOL_QUADROTOR	Quad-rotor VTOL using a V-shaped quad config in vertical operation. Tailsitter.
21	MAV_TYPE_VTOL_TILTROTOR	Tiltrotor VTOL

to fixed wing. 

Thanks for tracking this down! As I recall, back when the core code was written APM only used those two constants for all it's vehicles. We seem to be following the mavlink standard more closely these days. 

Anyway, for now just to confirm please replace IOBoard_Funcs.ino, line 119 with
`  if(apm_mav_type == 2 || apm_mav_type == 13) { // ArduCopter MultiRotor or ArduCopter Heli`
and see if that fixes it for you.  
Cheers,
Nick

**Comment by CerberusVelvet** [comment]
No worries! I will try and report back asap

**Comment by CerberusVelvet** [comment]
Yep, that did the trick!

I also added:
```
 out(F("Vehicle Type:"), stream);
 outln(apm_mav_type, stream); 
```

to `void dumpVars(Stream *stream)`, to confirm the vehicle type.


```
MavPixel 2.1 ready.
#vars

Sats:11
Fix:3
Hdop:0
Modes:5
**Vehicle Type:13**
Armed:0
Thr:0
BatVolt:14.47
```

Which confirms the type is 13. Now that that's corrected 'flMode' is reporting the correct value.

Cheers!
Cerbs

**Comment by prickle** [comment]
Added the set of constants as mentioned above for better vehicle type matching.

Thanks again Cerbs for sorting this out and testing the fix. I would love to see a picture of your rig all lit up!
Top job!
Nick.

**Comment by CerberusVelvet** [comment]
Excellent, happy to help!

Short vid of the first flight with the Mavpixel setup - https://www.youtube.com/watch?v=5RLfwFzAqRM

You can just about see the front arm neopixels changing colour with mode changes :)

Cerbs

**Comment by CerberusVelvet** [comment]
The more I watch it the more I think I want pixels under the middle arms of the hex. 

I am going to try adding two more strips. Wired in parallel with the A2 and A3 data pin. So they should be duplicates of 16-23 and 24-31

Totally off topic. Sorry!

Actually this issue can be closed now I guess? But I can't see how to do that :(

**Comment by prickle** [comment]
As far as I am concerned, more pixels = more goodness :-) All good Cerbs, I'll close it now.
Cheers!
Nick
